src/transformers/cache_utils.py

```diff
@@ -486,23 +501,37 @@ def crop(self, max_length: int):
                 self.key_cache[idx] = self.key_cache[idx][..., :max_length, :]
                 self.value_cache[idx] = self.value_cache[idx][..., :max_length, :]
 
-    def batch_split(self, full_batch_size: int, split_size: int, num_hidden_layers: int) -> List["DynamicCache"]:
+    def batch_split(
+        self, full_batch_size: int, split_size: int, num_hidden_layers: int = None
+    ) -> List["DynamicCache"]:
         """Split the current instance into a list of `DynamicCache` by the batch size. This will be used by
         `_split_model_inputs()` in `generation.utils`"""
+        if num_hidden_layers is not None:
+            warnings.warn(
+                "The `num_hidden_layers` argument is deprecated and will be removed in v4.47.0. There is no need to "
+                "pass this argument anymore.",
+                UserWarning,
+            )
         out = []
         for i in range(0, full_batch_size, split_size):
-            current_split = DynamicCache(num_hidden_layers)
+            current_split = DynamicCache()
             current_split._seen_tokens = self._seen_tokens
             current_split.key_cache = [tensor[i : i + split_size] for tensor in self.key_cache]
             current_split.value_cache = [tensor[i : i + split_size] for tensor in self.value_cache]
             out.append(current_split)
         return out
 
     @classmethod
-    def from_batch_splits(cls, splits: List["DynamicCache"], num_hidden_layers: int) -> "DynamicCache":
+    def from_batch_splits(cls, splits: List["DynamicCache"], num_hidden_layers: int = None) -> "DynamicCache":
         """This is the opposite of the above `batch_split()` method. This will be used by `stack_model_outputs` in
         `generation.utils`"""
-        cache = cls(num_hidden_layers)
+        if num_hidden_layers is not None:
+            warnings.warn(
+                "The `num_hidden_layers` argument is deprecated and will be removed in v4.47.0. There is no need to "
+                "pass this argument anymore.",
+                UserWarning,
+            )
```

Comment:
Let's warn once. We can also use the wrapper introduce by @qubvel : https://github.com/huggingface/transformers/blob/ce81eb8d14ae39e183a8956cce3f7e2e46ecaa8f/src/transformers/utils/deprecation.py#L32